### ISSUE TITLE
onupdate is triggered, even if field is explicitly set to previous value

### ISSUE THREAD

**Issue by Tohnmeister** [sqlalchemy/sqlalchemy#6142]
I have a model like the following:

```python
class Person(Base):
    __tablename__   = "person"

    id              = Column(Integer, primary_key=True, autoincrement=True)
    last_changed_by_script = Column(Boolean, default=False, onupdate=False, nullable=False)

   # ... some other fields
```

The  `last_changed_by_script` is a boolean column which should be set to FALSE if not explicitly set by code.

This works well if:

- The value was previously TRUE or FALSE and is untouched.
- The value was previously FALSE and is set to TRUE, e.g. `person.last_changed_by_script = True`.

But works incorrectly (or atleast different than I expected), if:

- The value was previously TRUE and I want it to stay TRUE by explicitly writing `person.last_changed_by_script = True`.

In the latter case, SQLAlchemy will consider the column to be unchanged, and hence will use the default (False).

The only way I managed to work around this, is by first setting the field to False, flushing and then setting it to True, before committing.

```python
person.last_changed_by_script = False
database_session.flush()
person.last_changed_by_script = True
database_session.commit()
```

Maybe this is more of a question than an issue. How can I force SQLAlchemy to stick to an existing value, even though the column is configured to use onupdate?

**Comment by zzzeek** [comment]
hi -

if the unit of work sees a column that has no net change, it does not include it in the UPDATE statement, that's not a bug.   Then i would assume you have other columns coming in, that *are* being UPDATEd, so the unit of work again correctly emits an UPDATE statement.  Then you have a rule on your Table to **uncondtionally** set it back to False if the column isn't changing, which is kind of an awkward pattern to use here.

I think the only way to force a value in there, is actually pretty easy, assign the SQL element true instead of the value true:

```
from sqlalchemy import true
object.my_attribute = true()
```

other ways that might be more robust and less mysterious would be to use something like a counter or a timestamp that is continuously changing.

**Comment by Tohnmeister** [comment]
@zzzeek thanks a lot for the explanation. I will go ahead and try your suggestion.

I understand the awkwardness of the pattern. Although the reasoning is completely unrelated to SQLAlchemy, let me explain a bit.

We have a ton of legacy code which already uses this `Person` model. Now, recently we've added some scripts that modify some Person records. But they should only do so if the `Person` was last modified or created by a script. Otherwise they should leave the `Person` alone. For that we've added the field `last_changed_by_script`. If the script creates the `Person` record the field is set to true. If a user modifies the `Person` record later through our web app, the field is set to false, and any script should leave it alone.

Now ,we could go ahead and modify the existing code (where `Person` records are modified/created by users) and explicitly set the field to false in all these cases. But with the risk of forgetting places now or in the future, we've decided that an alternative approach is to make the field false onupdate. And only set it to true explicitly in the newly added scripts.

As said, out of scope for SQLAlchemy. And most likely we'll revert this change and go back to making it explicit, because of the awkwardness of the pattern. I'll still go ahead and try your suggestion. Thanks a lot!
